### ISSUE TITLE
AddSystemsManager Takes 1+ Minutes to Complete

### ISSUE THREAD

**Issue by arex388** [aws/aws-dotnet-extensions-configuration#122]
### Describe the bug

I have a .NET 6 solution with several ASP.NET Core apps and supporting libraries. The ASP.NET Core apps all pull their configuration from AWS SSM Parameter Store. The problem that I'm running into is that it takes more than 1 minute for the process to complete.

I attempted to trace where the slow down is and it's coming from the `builder.Add(source);` call in the AddSystemsManager method with signature `AddSystemsManager(this IConfigurationBuilder builder, Action<SystemsManagerConfigurationSource> configureSource)`, [line 188](https://github.com/aws/aws-dotnet-extensions-configuration/blob/22d5c3e5efadfec360251bc98e5e02cc13a526ae/src/Amazon.Extensions.Configuration.SystemsManager/SystemsManagerExtensions.cs#L188).

I also attempted to trace into the `Add` method, but I was getting build errors with the .NET runtime source and couldn't.

So, something about how the source is being added and processed takes an excessively long time. This makes integration tests take significantly longer because they have to wait 1+ minutes for the app to spin up, just so they can do 5-10 sec of actual work. It also makes manual development take significantly longer because the app has to rebuild and reinitialize on each change. Ultimately it's very aggravating because I spend more time waiting for the app to spin up than I do developing or testing.

This is how the app behaves in development locally on my machine, in staging on an EC2 staging server, and in production also on an EC2 production server. In production it's especially bad because the apps take so long to spin up before they can accept traffic that we have no choice but to loose reports coming in from IoT devices while the app listening to them is initializing. We have ~2m IoT devices, so it's not good.

### Expected Behavior

I expect the process of adding the Parameter Store values to complete in a few seconds, not 1+ minutes.

### Current Behavior

The process takes 1+ minutes to add the Parameter Store values.

### Reproduction Steps

I'm sorry, it's simply impossible for me to create a repo for you. I've already spent enough time trying to trace it. All I can give you is the point where the slowdown occurs.

### Possible Solution

I don't know about a possible solution, but is it possible that the culprit is how long it takes for the credentials to be pulled? On my development machine I add credential profile using PowerShell, and I'm assuming on the EC2 instance it delegates the credentials somehow, but could the process of discovering those credentials be what is taking the bulk of the time? If it is, is it possible to make the discovery shorter by pointing it a specific credential source?

### AWS .NET SDK and/or Package version used

AWSSDK.SimpleSystemsManagement 3.7.100.6
Amazon.AspNetCore.DataProtection.SSM 3.0.0
Amazon.Extensions.Configuration.SystemsManager 4.0.0

### Targeted .NET Platform

.NET 6

### Operating System and version

Windows 10, Windows Server 2019, Windows Server 2022

**Comment by ashishdhingra** [comment]
Hi @arex388,

Good afternoon.

Could you please share the following:
- Does the delay occurs every time or only during the 1st request?
- Could you try temporarily enabling metrics logging and share more detailed logs? This could be enabled using the below statements (more details at [Retrieving Request Metrics from the AWS SDK for .NET](https://aws.amazon.com/blogs/developer/retrieving-request-metrics-from-the-aws-sdk-for-net/)):
```C#
AWSConfigs.LoggingConfig.LogResponses = Amazon.ResponseLoggingOption.Always;
AWSConfigs.LoggingConfig.LogMetrics = true;
AWSConfigs.LoggingConfig.LogTo = Amazon.LoggingOptions.SystemDiagnostics;
```

The 1st request could see some delay where credentials would need to be resolved in the configured profile chain (see [Credential and profile resolution](https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/creds-assign.html)).

Thanks,
Ashish

**Comment by arex388** [comment]
Hey @ashishdhingra 

I figured it out. When I started trying to capture the metrics as you requested I read through the output and noticed that even though I was trying to spin up the Development environment, the metrics were showing that parameters from the Staging and Production environments were also being loaded.

I was confused why that was, and after diging into the code for a bit I realized the issue... The problem was that I was calling `builder.Configuration.AddSystemsManager("/");` and the culprit specifically was the `"/"`. I guess my understanding of how configuration sources were loaded was wrong.

I thought the config wasn't being loaded until the following code ran, where I thought I was filtering down to the correct enviornment:

```
var sectionName = environment.EnvironmentName switch {
    "Development" => "config.development",
    "Staging" => "config.staging",
    _ => "config.release"
};
var section = configuration.GetSection(sectionName);
```

In reality what was happening was that SSM was downloading everything from the root path, and I was filtering it down to the correct environment after the fact. Generally I suppose it wouldn't be that much of an issue, but what slowed it down the most was the Data Protection keys that were being persisted in Parameter Store. There were thousands of them, which struck me as odd because I thought a key was supposed to last 1-3 months (going off memory). Turns out the production app was generating new keys every 3 days and persisting them. After 3 years of the app running that builds up and app initialization kept getting slower and slower. Perhaps since it was loading thousands of keys at some point it gives up and simply generates new keys which just adds fuel to the fire?

I did a test run and deleted all Data Protection keys for Development and Staging and the initialization reduced by 20 seconds, so it was now obvious what the issue was.

I have since refactored the configuration loading to call into specific environment paths instead of the root which acts as the natural filter I need: `builder.Configuration.AddSystemsManager($"/{builder.GetEnvironmentName()}/");`. And that simplifies the section loading to just `var section = configuration.GetSection("config");`. The new paths changed to `/{environment}/config`.

The end result from all of that is that app initialization times dropped from ~1.5min to 3-5sec, which is amazingly better. Working on the apps is now finally going to be fun (and speedy) again. Integration tests are quite speedy now too, no more waiting just for something to happen.

Thanks for pointing me to the metrics. Had you not done that I would have still been stuck with an ever slower and slower app initialization and the dread of working on the solution. Thanks again!
